### ISSUE TITLE
[regression - 5.4] Generated `.d.ts` file contains invalid types

### ISSUE THREAD

**Issue by bradzacher** [microsoft/TypeScript#57843]
### 🔎 Search Terms

error "generic type" "requires" "type argument(s)" declaration .d.ts emit

### 🕗 Version & Regression Information

- This changed between versions 5.3 and 5.4

### ⏯ Playground Link

[bug workbench link](https://www.typescriptlang.org/dev/bug-workbench/#code/PTAEAEGcAsHsHcCiBbAlgFwFAggEwKYDGANgIYBOp6qsAdgFyjrkCu+2YUcSa66+uAGKpi+RuXwAHcrAB0uWekiYOEAGYj8tUsjGgARqUj5Fy-AA9JsculCl9kZqUK2SRyKADytfAB5MoKAAIqAAvKAstADWtAi0ADQBoADKYRHRsfAJSQCyaZExcYmBQQAK+RlFScnl4QWZ2QB8oADeAL6YFlY2dg5OLqBukB4AKvCw-iVppLQAnsUp03MLeeEz80llSxuBNduJzRb8tLge3n5B8SlXOVdlVzXNLUkSpLh0xLN2AISMQQDcL3wbw+X30vxSgMCr3etE+gwhOShvUclAGMNBoFwjGSqHMvkuKUagI6nUs1lcZGGoBG0AkFyeHS6FMGVI8glgLHIBMZZO6tnQs0k+FAwgAbhcHs1wrT6TzQAAfUWc7nJZpKyIEDQ+XCA5k9eyo5yU9wpPEEiqFLIPS0NJ5AkFwr74CEAh2w+FqCHJEkqfUCoUi5L4CW0XxqtI+vksoYeRCoADm0HQBKloCOWlOZvx4slRIeIa04cavP9TEDoAAcqgfL4AMLp8zHLO48zS0DxpMpkZFuuNK6IUSh+slvXknqC4U03uN5seVvtusAbQARPgVwBdMf88tTweFsMNjMned4xertSb7csycisb4YgS3wjdvV2scrn8Isv4nRicV2lUBsfBv3bHsw3vR8-B-a8eljZVPxAiDDibTNT3xMVYFQXArhfVomXHWwa34cg1GcEVhDUL9kNaJIE0YEZfTLeDW2o59thQucaXGSYaWgIC2JfBYRgWSjBP7JIfAlcgFik-ByEwe1AlgaTyGwkUMSdLEIU7ZNfEw7DcN-Uky1vFID0E2c0O4rRBXOMDV2gK8-wDPdE2TJC0haUAlyiRhURrBMNwY+BbNmc5QDaWDXJFN98GoryfL80AAtoIKQtgSLot3O9QtoQU0gAJSIaxcF8VKEwePFqN-FzWVNe98vC2sRg42joWBD0vmgCFGMwAid3gxrBTGCZWuPLNhua-BOOsqbzl44NQ3ExV0i1GsBEU-C6qGvKRrpEDUyJKyT2zC0NROfBtQEG0LvWnVS0I+rqSmj9uUJZIbjuUoHlKWbTtG3iQjuq6Npw6pVs1UGdRWSHLuu8GSnKEGEYWPYUbBrbnmU1T1NATT4VwXq9tmWVDo+4y-SeszXtQJ9xtQgHxgcldcGcmmSdbXwADUTqzXTqKuQq+Y8OLBdAABBEWbKa9tsZ8gBpUAa1AKJ8FmWA1FAbngoQ4CiyW2ykJ5pcFY3Ud+tAAAyVpFeV2hVfVzXQEK3XaafQrTfNyLrfau2VbVjWtYlt2SdzXwJa94zso5sLDbDJJeYmuN3MFpJheTqsNrTwIpczqaDjSQCbFJg7oM581uaFq4JYt0yAJJgXeKTxn+dTpCFgz1vRezjukjz7uZcFQvwn3Ycpvjnnq8luvqYbsK4ub6WBb7wIu64sXV8l6WC62mVv05g8p5dmuLZp-iS-rZf260OWkl8+3HaD0A611t62Pjtjlyic3o8wc+BLTSvvnEmd9AgPwDk7LWr8QphXDt-c2JIY4AQviNHiR5B6b1vl5e+URH6B2djAoesx4G+UQf1bKQ1UGlzlMkYUhAd7ULev9Fs5okjF0FFzOhRAJKBA4SQumfhuGEEaFjAaN4UGALeolAmXxfL+WYIFXWrYsp1QASXcOQRYDIFIDWDw+cmEqjAfjTqmJTb4KgcEbRujaCQF1mPIsWidF6KjkgtRkiS5cycTY-Rg9+HMJwR1R08JzGQOft4vRG4AD8DFHHWJcWbP+9cpz8MngzLio0WaEGcsku81Cm7pLmszIuq59A5Lnik6hi9E7Xy7Fvde1ksHZFzjvEmI9bayP9g7AhWsdY4gssbbmUcfY23lp00J3TLGuwYtQyentEkkl9t5CZT9nYhxmYApukdEmqOYmyZUl8amZxXloTu0smkLAHhktpW0QETHYSTReVcT4z2Eo3VOx9Cqn2ElUjanzvmSULDJQF0ksZJBUvJNSBATHBK+ETDZJcybQWoTzZoNt+GA0KmivigCFq10pnVAgbgJCgDUJEFwNAHbvEAmlI6n1QC3GCD9FIf0AAUSQKD0W4hMD6X0mW-QkgASlgU1N6dK+X3BZbVVQ4ANCiG0LocQUgZCmEwKgZAO5vLUv4mlK4ZkOQlx9moGQyBQArlkMAQwxgVyULoI4AwFA0jasCqy9odh2QUlmL4doVwfX4UaIKwEMq5VaB0HoCQ0hYAAH1OToEkCwdA8hVVliJWQElhA7W2EMOQRg6r+SsvNZaow65BWyFeiqXieaKQFotVaktZbHl-LdTbKtNga1FutaWjFPF2jYt7QsVt6B211pXF295XZvVtF9oO4dxbR1luoYDXtSy2i8OVhq6thaR1dt+bWP1+7V0LHqFUQIx6siKUBEAA)

### 💻 Code

```ts
// @showEmit
// @declaration: true
// @showEmittedFile: repro.d.ts

// @filename: base.ts
export abstract class One<
  D = unknown,
  S = unknown,
  M = unknown,
  DP = unknown,
  SP = unknown,
> {}
export abstract class Two<
  D = any,
  S = any,
  M = any,
  DP = any,
  SP = any,
> extends One<D, S, M, DP, SP> {
  readonly a!: D;
  readonly b!: S;
  readonly c!: M;
  abstract readonly d: Six<D, S>;
}

export class Three<D> {}
export class Four<D> {}
export type Five<D, S> = Three<D> | Four<S> | undefined;
export abstract class Six<D = unknown, S = unknown> {
  readonly e!: D;
  readonly f!: S;
}

export type Seven<S> = S;
export class Eight<D, S> extends Six<Five<D, S>, Seven<S>> {}
export type Nine<C extends Six> = Eight<Ten<C>, Eleven<C>>;
export type Ten<C extends Six> = C["e"];
export type Eleven<C extends Six> = C["f"];
export type Twelve<T> = Nine<Fourteen<T>>;
export type Thirteen<T> = Ten<Twelve<T>>;
export class Fourteen<T> extends Six<void, T> {}
export interface Fifteen<T> {
  g: T;
}
export class Sixteen<T = any> extends Two<
  Thirteen<T>,
  T,
  Fifteen<T>,
  never,
  never
> {
  override readonly d!: Eight<void, T>;
}

export type Seventeen<T extends TwentyOne> = T["h"];
export type Eighteen = { [k: string]: TwentyOne };
export type Nineteen = { [k: string]: Two };
export type Twenty = Record<string, Sixteen>;

export class TwentyOne<T = any> {
  readonly h!: T;
}
export class TwentyTwo<T extends TwentyOne> extends TwentyOne<
  Seventeen<T> | undefined
> {}
export class TwentyThree<D, S> extends Six<D | undefined, S | undefined> {}
export class TwentyFour<D, S, M, DP, SP> extends Two<
  D | undefined,
  S | undefined,
  M | undefined,
  DP | undefined,
  SP | undefined
> {
  override readonly d!: TwentyThree<D, S>;
}

export type TwentyFive<T extends Two> = T["d"];
type TwentySix<V extends Eighteen, R extends Nineteen, A extends Twenty> = {
  [K in keyof V]: Fourteen<Seventeen<V[K]>>;
} & { [K in keyof R]: TwentyFive<R[K]> } & {
  [K in keyof A]: TwentyFive<A[K]>;
};
export type TwentySeven<
  V extends Eighteen,
  R extends Nineteen,
  A extends Twenty,
> = ThirtyThree<TwentySix<V, R, A>>;
export type TwentyEight<
  V extends Eighteen,
  R extends Nineteen,
  A extends Twenty,
> = Eleven<TwentySeven<V, R, A>>;
export type TwentyNine<
  V extends Eighteen,
  R extends Nineteen,
  A extends Twenty,
> = Ten<TwentySeven<V, R, A>>;
type Thirty<C extends Eighteen> = {
  [k in keyof C]: Fourteen<Seventeen<C[k]>>;
};
type ThirtyOne<C extends Twenty> = {
  [k in keyof C]: TwentyFive<C[k]>;
};
export type ThirtyTwo<C extends Nineteen> = {
  [k in keyof C]: TwentyFive<C[k]>;
};
export class ThirtyThree<Spec extends ThirtyFour> extends Six<
  ThirtySix<Spec>,
  ThirtyFive<Spec>
> {}
export type ThirtyFour = { readonly [k: string]: Six };
export type ThirtyFive<Domains extends ThirtyFour> = {
  readonly [K in keyof Domains]: Eleven<Domains[K]>;
};
export type ThirtySix<Domains extends ThirtyFour> = {
  readonly [K in keyof Domains]?: Ten<Domains[K]>;
};
export type ThirtySeven<T extends Two> = T["c"];
export type ThirtyEight<T extends Two> = T["b"];
export type ThirtyNine<
  V extends Eighteen,
  R extends Nineteen,
  A extends Twenty,
> = { readonly [K in keyof V]: Seventeen<V[K]> } & {
  readonly [K in keyof R]: ThirtySeven<R[K]>;
} & { [K in keyof A]: ThirtyEight<A[K]> };
export class Forty<
  V extends Eighteen,
  R extends Nineteen,
  A extends Twenty,
> extends Two<
  TwentyNine<V, R, A>,
  TwentyEight<V, R, A>,
  ThirtyNine<V, R, A>,
  never,
  never
> {
  override readonly d!: ThirtyThree<Thirty<V> & ThirtyTwo<R> & ThirtyOne<A>>;
}

export declare function doThing<D, S, M, DP, SP>(
  arg: Two<D, S, M, DP, SP>,
): TwentyFour<D, S, M, DP, SP>;


// @filename: repro.ts
import { doThing, type Forty } from "./base";
export const bar = doThing({} as Forty<{}, {}, {}>);

// @filename: repro_output.d.ts
export declare const bar: import("./base").TwentyFour<
  import("./base").TwentyNine<{} & import("./base").ThirtyTwo<{}> & {}>,
  import("./base").TwentyEight<{} & import("./base").ThirtyTwo<{}> & {}>,
  import("./base").ThirtyNine<{}, {}, {}>,
  unknown,
  unknown
>;
```

### 🙁 Actual behavior

The code emitted for `repro.d.ts` is
```ts
export declare const bar: import("./base").TwentyFour<
  import("./base").TwentyNine<
    import("./base").Thirty<{}> &
      import("./base").ThirtyTwo<{}> &
      import("./base").ThirtyOne<{}>
  >,
  import("./base").TwentyEight<
    import("./base").Thirty<{}> &
      import("./base").ThirtyTwo<{}> &
      import("./base").ThirtyOne<{}>
  >,
  import("./base").ThirtyNine<{}, {}, {}>,
  unknown,
  unknown
>;

```

This code has two type errors (on lines 2-6 and 7-11 respectively):
```
Generic type 'TwentyNine' requires 3 type argument(s).
Generic type 'TwentyEight' requires 3 type argument(s).
```

Note how the generated type uses `TwentyNine` on L2 and `TwentyEight` on L7.
These are the incorrect types to use (see below)

### 🙂 Expected behavior

The types generate valid code with no semantic errors.
For example intellisense reports this type for `bar` which is valid
```ts
export declare const bar: import("./base").TwentyFour<
  import("./base").ThirtySix<
    import("./base").Thirty<{}> &
      import("./base").ThirtyTwo<{}> &
      import("./base").ThirtyOne<{}>
  >,
  import("./base").ThirtyFive<
    import("./base").Thirty<{}> &
      import("./base").ThirtyTwo<{}> &
      import("./base").ThirtyOne<{}>
  >,
  import("./base").ThirtyNine<{}, {}, {}>,
  unknown,
  unknown
>;
```

Note how this type instead uses `ThirtySix` on L2 and `ThirtyFive` on L7.

### Additional information about the issue

Sorry that this example is so goddamn disgusting to look at.
This is the name-mangled version of some real code from our codebase.
I tried to minimise it but it's all such an intermingled spaghetti mess that I wasn't able to figure out what things I could delete without changing the output.

For context - I am working on upgrading our codebase to TS5.4.
When I ran our typecheck CLI I got a number of errors across the codebase.
When I opened the files with errors - those errors didn't show up in the IDE.
I spent a while pulling my hair out trying to figure out why there was a discrepancy.

The error messages led me back to a monster file which makes use of some really ugly patterns of inferred types from `typeof`s to generate a lot of types. For reference the file itself is >3k LOC and the `.d.ts` file it generates is >11k LOC. So so so many anonymous, inferred types.

When I opened the `.d.ts` file it was filled with type errors. From what I can tell what's happening is that when you open the files with errors in the IDE then TS uses the `.ts` source files for type checking and so it uses the "correct" type for everything. When we run our CLI to do the typecheck it uses project references and so it uses the `.d.ts` output instead.

Because the `.d.ts` output doesn't match the types TS infers from the `.ts` file this causes the discrepancy between the reported errors.

I think a most (if not all) of the errors in the `.d.ts` file are variations of the error shown by this repro but it's so hard to know as there are hundreds of them.

Note: I didn't write this code and I hate that it exists. I am sorry.

**Comment by bradzacher** [comment]
Update: This is a regression in TS5.4

In TS5.3 the emitted `.d.ts` file is correct, but in TS5.4 it uses the wrong aliases

**Comment by bradzacher** [comment]
I've got a repro repo for testing too.
1) clone https://github.com/bradzacher/bug-repros
2) checkout the `typescript-dts-emit-bug` branch
3) `npm i`
4) `npm test` --> observe diff

Bisecting it looks like the first release that exhibited this bug was `5.4.0-dev.20240110`

**Comment by jakebailey** [comment]
If you hadn't tried it yet, https://www.npmjs.com/package/every-ts can bisect further down from the nightly versions.

**Comment by bradzacher** [comment]
Bisecting with `every-ts` reveals 4bcbc16cff6c65caf349f1c1df4dc7b6809468ae to be the bad commit

**Comment by bradzacher** [comment]
Yup - 100% confirmed.
4bcbc16cff6c65caf349f1c1df4dc7b6809468ae only has 4 lines of changes so it's easy to revert locally. If I comment out the 4 lines then it emits the correct output.

**Comment by bradzacher** [comment]
I patched #57396 locally and it looks like it fixes the regression
